### PR TITLE
chunk: add Iterator interface.

### PR DESCRIPTION
**Pull request by coocood**
This abstraction makes algorithms like join result generator able to iterator chunk rows in different forms.

### CODE REVIEW

File: util/chunk/iterator.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+type sliceIterator struct {
+	rows   []Row
+	cursor int
+}
+
+func (it *sliceIterator) Begin() Row {
+	if len(it.rows) == 0 {
+		return Row{}
+	}
+	it.cursor = 1
```
Comment (zz-jason) [inline util/chunk/iterator.go]: let `it.cursor` point at the current row may be more readable.

Comment (coocood) [reply util/chunk/iterator.go]: After `Begin` is called, the current row is the row at index 1.

Comment (coocood) [reply util/chunk/iterator.go]: It is equivalent to
```go
row := it.rows[0]
it.cursor = 1
return row
```

Comment (zz-jason) [reply util/chunk/iterator.go]: I mean we can let `e.cursor` point to the row idx of the returned `row`.

Comment (coocood) [reply util/chunk/iterator.go]:
```go
func (it *listIterator) Next() Row {
	if it.chkCursor == it.li.NumChunks() {
		return Row{}
	}
	chk := it.li.GetChunk(it.chkCursor)
	if it.rowCursor + 1 == chk.NumRows() {
		it.chkCursor++
		it.rowCursor = 0
		if it.chkCursor == it.li.NumChunks() {
			return Row{}
		}
	} else {
		it.rowCursor++
	}
	if it.chkCursor == it.li.NumChunks() {
		return Row{}
	}
	chk = it.li.GetChunk(it.chkCursor)
	return chk.GetRow(it.rowCursor)
}
```

Here is the code of `listIterator` Next when cursor points to the returned Row, I think it's harder to read than the current implementation.

```go
func (it *listIterator) Next() Row {
	if it.chkCursor == it.li.NumChunks() {
		return Row{}
	}
	chk := it.li.GetChunk(it.chkCursor)
	row := chk.GetRow(it.rowCursor)
	it.rowCursor++
	if it.rowCursor == chk.NumRows() {
		it.rowCursor = 0
		it.chkCursor++
	}
	return row
}
```

Comment (zz-jason) [reply util/chunk/iterator.go]: ok

File: util/chunk/iterator.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+type sliceIterator struct {
+	rows   []Row
+	cursor int
+}
+
+func (it *sliceIterator) Begin() Row {
+	if len(it.rows) == 0 {
+		return Row{}
```
Comment (zz-jason) [inline util/chunk/iterator.go]: how about return `it.End()` once we reach the end of iterator during the execution of `Begin()` and `Next()`

**Pull request merged**